software/firmware/europi.py

```diff
@@ -1,249 +1,278 @@
-        else:
-            self.off()
+# Default font is 8x8 pixel monospaced font.
+CHAR_WIDTH = 8
+CHAR_HEIGHT = 8
 
 
+# Helper functions.
 
+def clamp(value, low, high):
+    return max(min(value, high), low)
 
-class AnalogueInput: #Class used to read the analogue input
-    def __init__(self, pin):
-        self.input = ADC(Pin(pin))
-        self.input_multiplier, self.input_offset = get_input_calibration_data() #Retreives the calibration data from the calibration.txt file
 
-    def read_duty(self, samples=256): #Reads the un-corrected duty cycle of the ADC
-        return clamp(sample_adc(self.input, samples), 0, MAX_UINT16)
+def reset_state():
+    """Return device to initial state with all components off and handlers reset."""
+    oled.clear()
+    [cv.off() for cv in cvs]
+    [d.reset_handler() for d in (b1, b2, din)]
 
-    def read_voltage(self, samples=256): #Reads the voltage read by the ADC, corrected to convert both duty to voltage, and also correct the offset
-        return clamp((self.read_duty(samples) * self.input_multiplier) + self.input_offset, 0, 12)
 
+# Component classes.
 
+class AnalogueReader:
+    def __init__(self, pin, samples=DEFAULT_SAMPLES):
+        self.pin = ADC(Pin(pin))
+        self.set_samples(samples)
 
+    def _sample_adc(self, samples=None):
+        # Over-samples the ADC and returns the average.
+        values = []
+        for _ in range(samples or self._samples):
+            values.append(self.pin.read_u16())
+        return round(sum(values) / len(values))
 
-class Knob: #Class used to read the knob positions
-    def __init__(self, pin):
-        self.input = ADC(Pin(pin)) #The knobs are 'read' by analogue to digital converters
+    def set_samples(self, samples):
+        """Override the default number of sample reads with the given value."""
+        if not isinstance(samples, int):
+            raise ValueError(
+                f"set_samples expects an int value, got: {samples}")
+        self._samples = samples
 
-    def unit_interval(self, samples=256):
-        return 1 - (sample_adc(self.input, samples) / MAX_UINT12) #Provide the knob's position as a float value between 0.0 and 1.0
+    def percent(self, samples=None):
+        """Return the percentage of the component's current relative range."""
+        return self._sample_adc(samples) / MAX_UINT16
 
-    def read_position(self, steps=100, samples=256): #Returns an int in the range of steps based on knob position.
+    def range(self, steps=100, samples=None):
+        """Return a value from steps chosen by the current voltages relative position."""
         if not isinstance(steps, int):
-            raise Exception("Please only use integer type for steps and decimal_places with the read_position method")
+            raise ValueError(f"range expects an int value, got: {steps}")
+        percent = self.percent(samples)
+        if percent == 1.0:
+            return steps
+        return int(self.percent(samples) * steps)

```

Comment:
is it necessary to call `percent()` a second time here?  And if not is the 1.0 special case worth handling?

Reply:
Ah! Good catch, that's my bad. No need for the second call to percent. Also, good catch with the special case. Should `range(steps)` be inclusive of the upper bound value of steps? For example, should `steps(100)` return 0 to 99 or 0 to 100?

Reply:
I'm a little split on this. My programmer mind says that 0-99 is 'right'. The user asked for 100 steps and that's what they should get, and in general ranges are non-inclusive. However, my UX mind says that the user would probably expect a knob that goes to 100 if they called this.

I think that as long as we document out well either choice is fine. I'm leaning towards 0-99 though.